flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveDynamicTableFactory.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connectors.hive;
+
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.catalog.CatalogTable;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.factories.DynamicTableSinkFactory;
+import org.apache.flink.table.factories.DynamicTableSourceFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.flink.table.catalog.config.CatalogConfig.IS_GENERIC;
+
+/**
+ * A dynamic table factory implementation for Hive catalog. Now it only support generic tables.
+ * Hive tables should be resolved by {@link HiveTableFactory}.
+ */
+public class HiveDynamicTableFactory implements
+		DynamicTableSourceFactory,
+		DynamicTableSinkFactory {
+
+	@Override
+	public String factoryIdentifier() {
+		throw new UnsupportedOperationException("Hive factory is only work for catalog.");
+	}
+
+	@Override
+	public Set<ConfigOption<?>> requiredOptions() {
+		throw new UnsupportedOperationException("Hive factory is only work for catalog.");
+	}
+
+	@Override
+	public Set<ConfigOption<?>> optionalOptions() {
+		throw new UnsupportedOperationException("Hive factory is only work for catalog.");
+	}
+
+	private static CatalogTable removeIsGenericFlag(CatalogTable table) {
+		Map<String, String> newOptions = new HashMap<>(table.getOptions());
+		boolean isGeneric = Boolean.parseBoolean(newOptions.remove(IS_GENERIC));
```

Comment:
Shouldn't this issue be solved earlier? Actually, the Hive CatalogTable should already remove this property and return it via the `isGeneric` method.

Reply:
You mean when take table out of the catalog? But `HiveFactory` can not know whether it's a generic table or a hive table.
Now the `CatalogTable` of hive table is not a standard Flink `CatalogTable`, it dose not have legacy "connector.type" or "connector" property. But I think we can make it Flink style.

Reply:
From a logical perspective, the `HiveCatalog` should deal with this additional property while storing and retrieving a `CatalogTable`. The `isGeneric` should not travel through the stack of regular Flink connector options. It is an internal property from the `HiveCatalog` through a custom `HiveCatalogTable` to a custom `HiveFactory`. It should be stored as a member variable in a `HiveCatalogTable`.

Reply:
Yes, this is our plan. Whether it's `createTable` or `getTable` of `HiveCatalog`, there should be two types of objects that are consistent: `HiveCatalogTable` and Flink `CatalogTableImpl`.
But `HiveCatalogTable` should implement `toProperties` too, it can add a `"connector"="hive"`.
And I think it is not a small change, we can finish this refactoring in 1.12.

Now the underlying rules are:
- HiveCatalog.createTable , hive table has false of `isGeneric` flag, generic table has no flag.
- HiveCatalog.getTable , generic table has true of `isGeneric` flag, hive table has no flag.
It is very difficult to understand.

Reply:
JIRA: https://issues.apache.org/jira/browse/FLINK-17922